File: bin/bamboo/dht/gateway_test.c
/*
 * Copyright (c) 2001-2003 Regents of the University of California.
 * All rights reserved.
 *
 * See the file LICENSE included in this distribution for details.
 */

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include "gateway_prot.h"

#define APP_STRING "gateway_test.c $Revision: 1.9 $"
#define CLIB_STRING "rpcgen"

static void 
do_null_call (CLIENT *clnt) {
    char *null_args = NULL; 
    void *null_result; 
    printf ("Doing a null call.\n");
    null_result = bamboo_dht_proc_null_2((void*)&null_args, clnt);
    if (null_result == (void *) NULL) {
        clnt_perror (clnt, "null call failed.");
        exit (1);
    }
    printf ("Null call successful.\n");
}

static void
do_put (CLIENT *clnt, bamboo_put_args *put_args) {

    bamboo_stat     *put_result;

    printf ("Doing a put\n");

    put_args->application = APP_STRING;
    put_args->client_library = CLIB_STRING;
    put_args->ttl_sec = 3600; /* 1 hour */

    put_result = bamboo_dht_proc_put_2 (put_args, clnt);
    if (put_result == (bamboo_stat *) NULL) {
        clnt_perror (clnt, "put failed");
        exit (1);
    }

    printf ("Put successful\n");
}

static bamboo_get_res*
do_get (CLIENT *clnt, bamboo_get_args *get_args) {
    bamboo_get_res  *get_result;

    printf ("Doing a get\n");

    get_args->application = APP_STRING;
    get_args->client_library = CLIB_STRING;
    get_args->maxvals = 1;

    get_result = bamboo_dht_proc_get_2 (get_args, clnt);
    if (get_result == (bamboo_get_res *) NULL) {
        clnt_perror (clnt, "get failed");
        exit (1);
    }
    else {
        return get_result;
    }
}

static CLIENT*
connect_to_server (struct sockaddr_in *addr) {
    CLIENT *clnt;
    int sockp = RPC_ANYSOCK;
    clnt = clnttcp_create (addr, BAMBOO_DHT_GATEWAY_PROGRAM, 
            BAMBOO_DHT_GATEWAY_VERSION, &sockp, 0, 0);
    if (clnt == NULL) {
        clnt_pcreateerror ("create error");
        exit (1);
    }
    return clnt;
}

static int 
compare_values (bamboo_value *val, char* v2, int l2) {
    int i;
    if (val->bamboo_value_len != l2)
        return val->bamboo_value_len - l2;
    for (i = 0; i < val->bamboo_value_len; ++i) {
        if (val->bamboo_value_val[i] != v2[i])
            return val->bamboo_value_val[i] - v2[i];
    }
    return 0;
}

static void 
random_key (char *key_val, int key_len) {
    int i;
    key_val [0] = 0;  // make sure positive
    for (i = 1; i < key_len; ++i)
        key_val [i] = rand ();
}

static void
run_test (struct sockaddr_in *addr)
{
    CLIENT *clnt;
    char value_val1[] = "Hello, world.";
    char value_val2[] = "Goodbye, world.";
    bamboo_put_args put_args;
    bamboo_get_args get_args;
    bamboo_get_res  *get_result;
    int first;

    memset (&put_args, 0, sizeof (put_args));
    memset (&get_args, 0, sizeof (get_args));

    srand (1);

    clnt = connect_to_server (addr);
    do_null_call (clnt);

    // Do a first put.

    random_key (put_args.key, sizeof (put_args.key));
    put_args.value.bamboo_value_val = value_val1;
    put_args.value.bamboo_value_len = sizeof (value_val1);
    do_put (clnt, &put_args);

    // Check that the data's there.

    memcpy (get_args.key, put_args.key, sizeof (get_args.key));
    get_result = do_get (clnt, &get_args);

    if (get_result->values.values_len != 1) {
        printf ("Get failed: returned %d values.\n", 
                get_result->values.values_len);
        exit (1);
    }
    if (compare_values (&(get_result->values.values_val [0]),
             value_val1, sizeof (value_val1)) != 0) {
        printf ("Get failed: values don't match: %s vs %s\n", 
                value_val1,
                get_result->values.values_val [0].bamboo_value_val);
        exit (1);
    }
    printf ("Get successful.\n");

    // Do a second put with the same key.

    put_args.value.bamboo_value_val = value_val2;
    put_args.value.bamboo_value_len = sizeof (value_val2);
    do_put (clnt, &put_args);

    // Check that both values are there.

    get_result = do_get (clnt, &get_args); 
    if (get_result->values.values_len != 1) {
        printf ("Get failed: returned %d values.\n", 
                get_result->values.values_len);
        exit (1);
    }

    printf ("Get returned value %s.\n", 
            get_result->values.values_val [0].bamboo_value_val);

    if (compare_values (&(get_result->values.values_val [0]),
             value_val1, sizeof (value_val1)) == 0) {
        printf ("Get returned first value.\n");
        first = TRUE;
    }
    else if (compare_values (&(get_result->values.values_val [0]),
                    value_val2, sizeof (value_val2)) == 0) {
        printf ("Get second first value.\n");
        first = FALSE;
    }
    else {
        printf ("Get failed: returned neither value.\n");
        exit (1);
    }

    get_args.placemark.bamboo_placemark_val = 
        get_result->placemark.bamboo_placemark_val;
    get_args.placemark.bamboo_placemark_len = 
        get_result->placemark.bamboo_placemark_len;

    get_result = do_get (clnt, &get_args); 
    if (get_result->values.values_len != 1) {
        printf ("Get failed: returned %d values.\n", 
                get_result->values.values_len);
        exit (1);
    }

    printf ("Get returned value %s.\n", 
            get_result->values.values_val [0].bamboo_value_val);

    if (first) {
        if (compare_values (&(get_result->values.values_val [0]),
                value_val2, sizeof (value_val2)) != 0) {
        printf ("Get failed: second value doesn't match: %s vs %s\n", 
                value_val2,
                get_result->values.values_val [0].bamboo_value_val);
        exit (1);
        }
    }
    else if (compare_values (&(get_result->values.values_val [0]),
                value_val1, sizeof (value_val1)) != 0) {
        printf ("Get failed: second value doesn't match: %s vs %s\n", 
                value_val1,
                get_result->values.values_val [0].bamboo_value_val);
        exit (1);
    }

    printf ("Get successful.\n");

    // Do a put with a different key.

    random_key (put_args.key, sizeof (put_args.key));
    do_put (clnt, &put_args);

    // Check that the data's there.

    memcpy (get_args.key, put_args.key, sizeof (get_args.key));
    get_args.placemark.bamboo_placemark_val = NULL;
    get_args.placemark.bamboo_placemark_len = 0;
    get_result = do_get (clnt, &get_args);

    if (get_result->values.values_len != 1) {
        printf ("Get failed: returned %d values.\n", 
                get_result->values.values_len);
        exit (1);
    }
    if (compare_values (&(get_result->values.values_val [0]),
             value_val2, sizeof (value_val2)) != 0) {
        printf ("Get failed: values don't match: %s vs %s\n", 
                value_val2,
                get_result->values.values_val [0].bamboo_value_val);
        exit (1);
    }
    printf ("Get successful.\n");

    clnt_destroy (clnt);
}

static void
lookup_server (char *host, int port, struct sockaddr_in *addr) {

    struct hostent *h;
    h = gethostbyname (host); 
    if (h == NULL) {
        perror ("gethostbyname");
        exit (1);
    }

    bzero (addr, sizeof (struct sockaddr_in));
    addr->sin_family = AF_INET;
    addr->sin_port = htons (port);
    addr->sin_addr = *((struct in_addr *) h->h_addr);
}

int
main (int argc, char *argv[])
{
    struct sockaddr_in addr;

    if (argc < 2) {
        printf ("usage: %s server_host server_port\n", argv[0]);
        exit (1);
    }

    lookup_server (argv [1], atoi (argv[2]), &addr);
    run_test (&addr);
    exit (0);
}

File: src/bamboo/dht/GatewayTest.java
/*
 * Copyright (c) 2001-2003 Regents of the University of California.
 * All rights reserved.
 *
 * See the file LICENSE included in this distribution for details.
 */

package bamboo.dht;
import java.net.InetAddress;
import java.security.MessageDigest;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Random;
import java.util.Set;
import org.acplt.oncrpc.OncRpcProtocols;
import org.apache.log4j.ConsoleAppender;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.log4j.PatternLayout;
import ostore.util.ByteUtils; // for print_bytes function only

/**
 * A simple test.
 *
 * @author Sean C. Rhea
 * @version $Id: GatewayTest.java,v 1.17 2004/11/13 20:34:23 srhea Exp $
 */
public class GatewayTest {

    public static class PutInfo {
        public byte [] key;
        public long expiry_time_ms;
        public PutInfo (byte [] k, long e) { key = k; expiry_time_ms = e; }
        public int hashCode () {
            int result = 0;
            for (int i = 0; i < key.length; ++i)
                result = result * 7 + (((int) key [i]) & 0xff);
            return result;
        }
        public boolean equals (Object rhs) {
            PutInfo other = (PutInfo) rhs;
            if (other.key.length != key.length)
                return false;
            for (int i = 0; i < key.length; ++i)
                if (key [i] != other.key [i])
                    return false;
            return true;
        }
    }

    public static class ClientThread extends Thread {

        protected Random rand;
        protected Set puts = new HashSet ();
        protected InetAddress gateway_host;
        protected int gateway_port;
        protected Logger logger;
        protected MessageDigest digest;
        protected gateway_protClient client;

        public ClientThread (InetAddress h, int p, int seed) {
            rand = new Random (seed); gateway_host = h; gateway_port = p;
            logger = Logger.getLogger (ClientThread.class);
            try { digest = MessageDigest.getInstance("SHA"); }
            catch (Exception e) { assert false; }
        }

        public void run () {

            // Connect to the gateway.

            try {
                sleep (rand.nextInt (5000));
            }
            catch (InterruptedException e) {}

            logger.info ("Connecting to host.");

            try {
                client = new gateway_protClient (gateway_host,
                        gateway_port, OncRpcProtocols.ONCRPC_TCP);
            }
            catch (Exception e) {
                logger.fatal ("Couldn't connect", e);
                System.exit (1);
            }

            while (true) {
                double which = rand.nextDouble ();

                // Make puts fall off as there are more and more values stored.
                if (which <= (1.0 / (puts.size ()/20 + 1))) {
                    // Do a put.

                    bamboo_put_args put = new bamboo_put_args ();

                    int size = rand.nextInt (1005) + 20;
                    put.application = "bamboo.dht.GatewayTest $Revision: 1.17 $";
                    put.client_library = "oncrpc";
                    put.value = new bamboo_value ();
                    put.value.value = new byte [size];
                    rand.nextBytes (put.value.value);
                    put.key = new bamboo_key ();
                    put.key.value = digest.digest (put.value.value);
                    put.ttl_sec = (rand.nextInt (10) + 1)*60;
                    int put_res = 0;
                    long start_time_ms = System.currentTimeMillis ();
                    try {
                        put_res = client.BAMBOO_DHT_PROC_PUT_2 (put);
                    }
                    catch (Exception e) {
                        logger.fatal ("Couldn't put", e);
                        System.exit (1);
                    }
                    long latency_ms =
                        System.currentTimeMillis () - start_time_ms;
                    if (put_res != bamboo_stat.BAMBOO_OK) {
                        logger.info ("Put failed (code=" + put_res
                                + "): size="
                                + put.value.value.length + " key="
                                + ByteUtils.print_bytes (put.key.value)
                                + " ttl_sec=" + put.ttl_sec + " lat="
                                + latency_ms + " ms");
                    }
                    else {
                        logger.info("Put successful: size="
                                    + put.value.value.length + " key="
                                    + ByteUtils.print_bytes(put.key.value, 0, 4)
                                    + " ttl_sec=" + put.ttl_sec + " lat="
                                    + latency_ms + " ms");

                        PutInfo pi = new PutInfo(put.key.value,
                                                 System.currentTimeMillis() +
                                                 put.ttl_sec * 1000);

                        assert!puts.contains(pi);
                        puts.add(pi);
                    }
                }
                else {
                    // Do a get.

                    // Choose a key.

                    PutInfo pi = null;
                    while (true) {

                        if (puts.size () == 0)
                            break;

                        int which_one = rand.nextInt (puts.size ());
                        Iterator i = puts.iterator ();
                        while (i.hasNext ()) {
                            pi = (PutInfo) i.next ();

                            // Don't look for things that are about to expire.
                            if (pi.expiry_time_ms <
                                    System.currentTimeMillis () + 30*1000) {
                                i.remove ();
                                pi = null;
                                continue;
                            }
                            else if (which_one-- == 0) {
                                break;
                            }
                        }

                        if (pi != null)
                            break;
                    }

                    if (pi != null) {

                        bamboo_get_args get_args = new bamboo_get_args ();
                        get_args.application = 
                            "bamboo.dht.GatewayTest $Revision: 1.17 $";
                        get_args.client_library = "oncrpc";
                        get_args.key = new bamboo_key ();
                        get_args.key.value = pi.key;
                        get_args.maxvals = 1;
                        get_args.placemark = new bamboo_placemark ();
                        get_args.placemark.value = new byte [] {};

                        long now_ms = System.currentTimeMillis ();

                        // Try for 20 seconds to retrieve it.

                        while (true) {

                            bamboo_get_res get_res = null;
                            try {
                                get_res =
                                    client.BAMBOO_DHT_PROC_GET_2 (get_args);
                            }
                            catch (Exception e) {
                                logger.fatal ("Couldn't get", e);
                                System.exit (1);
                            }
                            long latency_ms =
                                System.currentTimeMillis () - now_ms;

                            if (get_res.values.length == 0) {
                                logger.info ("Get failed: key=" +
                                        ByteUtils.print_bytes (pi.key, 0, 4));
                                if (System.currentTimeMillis ()
                                        < now_ms + 20*1000) {
                                    logger.info ("trying again");
                                }
                                else {
                                    logger.info ("giving up, lat="
                                            + latency_ms + " ms");
                                    break;
                                }
                            }
                            else {

                                PutInfo pi2 = new PutInfo (digest.digest (
                                            get_res.values [0].value), 0);
                                if (! pi.equals (pi2)) {
                                    logger.info ("Get got bad data: key="
                                            + ByteUtils.print_bytes (pi.key)
                                            + " value hash="
                                            + ByteUtils.print_bytes (pi2.key)
                                            + " lat="
                                            + latency_ms + " ms");
                                }
                                else {
                                    logger.info ("Get successful: key="
                                            + ByteUtils.print_bytes (
                                                pi.key, 0, 4) + " lat="
                                            + latency_ms + " ms");
                                }

                                break;
                            }


                            try {
                                sleep (1000);
                            }
                            catch (InterruptedException e) {}
                        }
                    }
                }

                long sleep_time = rand.nextInt (9001) + 1000;
                try {
                    sleep (sleep_time);
                }
                catch (InterruptedException e) {}
            }
        }
    }


    public static void main (String [] args) throws Exception {

        PatternLayout pl = new PatternLayout ("%d{ISO8601} %-5p %c: %m\n");
        ConsoleAppender ca = new ConsoleAppender (pl);
        Logger.getRoot ().addAppender (ca);
        Logger.getRoot ().setLevel (Level.INFO);

        if (args.length != 4) {
            Logger.getLogger (GatewayTest.class).fatal (
                    "usage: <gateway host> <gateway port> <clients> <seed>");
            System.exit (1);
        }

        InetAddress host = InetAddress.getByName (args [0]);
        int port = Integer.parseInt (args [1]);
        int clients = Integer.parseInt (args [2]);
        int seed = Integer.parseInt (args [3]);

        for (int i = 0; i < clients; ++i) {
            ClientThread t = new ClientThread (host, port, ++seed);
            t.start ();
        }
    }
}

